internal/services/containers/container_registry_cache_rule_resource.go

```diff
@@ -0,0 +1,277 @@
+		},
+
+		"target_repo": {
+			Type:        pluginsdk.TypeString,
+			Required:    true,
+			ForceNew:    true,
+			Description: "The target repository namespace such as 'ubuntu'.",
+		},
+	}
+}
+
+func (ContainerRegistryCacheRule) Attributes() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{}
+}
+
+type ContainerRegistryCacheRuleModel struct {
+	Name                string `tfschema:"name"`
+	ContainerRegistryId string `tfschema:"container_registry_id"`
+	CredentialSetId     string `tfschema:"credential_set_id"`
+	SourceRepo          string `tfschema:"source_repo"`
+	TargetRepo          string `tfschema:"target_repo"`
+}
+
+func (ContainerRegistryCacheRule) ModelObject() interface{} {
+	return &ContainerRegistryCacheRuleModel{}
+}
+
+func (ContainerRegistryCacheRule) ResourceType() string {
+	return "azurerm_container_registry_cache_rule"
+}
+
+func (r ContainerRegistryCacheRule) Create() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			cacheRulesClient := metadata.Client.Containers.CacheRulesClient
+			subscriptionId := metadata.Client.Account.SubscriptionId
+			ctx, cancel := timeouts.ForCreate(metadata.Client.StopContext, metadata.ResourceData)
+
+			var config ContainerRegistryCacheRuleModel
+			if err := metadata.Decode(&config); err != nil {
+				return err
+			}
+
+			defer cancel()
+			log.Printf("[INFO] preparing arguments for Container Registry Cache Rule creation.")
+
+			registryId, err := registries.ParseRegistryID(metadata.ResourceData.Get("container_registry_id").(string))
+			if err != nil {
+				return err
+			}
+
+			id := cacherules.NewCacheRuleID(subscriptionId,
+				registryId.ResourceGroupName,
+				registryId.RegistryName,
+				metadata.ResourceData.Get("name").(string),
+			)
+
+			existing, err := cacheRulesClient.Get(ctx, id)
+			if err != nil {
+				if !response.WasNotFound(existing.HttpResponse) {
+					return fmt.Errorf("checking for presence of existing %s: %s", id, err)
+				}
+			}
+
+			if !response.WasNotFound(existing.HttpResponse) {
+				return tf.ImportAsExistsError("azurerm_container_registry_cache_rule", id.ID())
+			}
+
+			// TODO: make a check that the repo is available in the registry.
+			// TODO: validate the source repo.
+
+			parameters := cacherules.CacheRule{
+				Name: &id.CacheRuleName,
+				Properties: &cacherules.CacheRuleProperties{
+					SourceRepository: pointer.To(config.SourceRepo),
+					TargetRepository: pointer.To(config.TargetRepo),
+				},
+			}
+
+			// Conditionally add CredentialSetResourceId if credentialSetId is not empty
+			if config.CredentialSetId != "" {
+				parameters.Properties.CredentialSetResourceId = pointer.To(config.CredentialSetId)
+			}
+
+			if err := cacheRulesClient.CreateThenPoll(ctx, id, parameters); err != nil {
+				return fmt.Errorf("creating %s: %+v", id, err)
+			}
+
+			metadata.SetID(id)
+
+			return metadata.Encode(&config)
```

Comment:
Wasn't quite sure if this function should return nil, and if not, if we should set properties of the config.

Reply:
Yep we return `nil` here, apologies if that was a misleading comment from me.
```suggestion
			return nil
```